packages/flutter/test/tracing/image_cache_tracing_test.dart

```diff
@@ -0,0 +1,95 @@
+// Copyright 2014 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:convert';
+import 'dart:developer' as developer;
+import 'dart:isolate' as isolate;
+
+import 'package:flutter/painting.dart';
+import 'package:vm_service/vm_service.dart';
+import 'package:vm_service/vm_service_io.dart';
+
+import '../flutter_test_alternative.dart';
+import '../painting/mocks_for_image_cache.dart';
+import '../rendering/rendering_tester.dart';
+
+void main() {
+  VmService vmService;
+  String isolateId;
+  setUpAll(() async {
+    final developer.ServiceProtocolInfo info =
+        await developer.Service.getInfo();
+    vmService = await vmServiceConnectUri(
+        'ws://localhost:${info.serverUri.port}${info.serverUri.path}ws');
+    await vmService.setVMTimelineFlags(<String>['Dart']);
+    isolateId = developer.Service.getIsolateID(isolate.Isolate.current);
+    TestRenderingFlutterBinding();
+  });
+
+  test('Image cache tracing', () async {
+    final TestImageStreamCompleter completer1 = TestImageStreamCompleter();
+    PaintingBinding.instance.imageCache.putIfAbsent(
+      'Test',
+      () => completer1,
+    );
+    PaintingBinding.instance.imageCache.clear();
+    final Timeline timeline = await vmService.getVMTimeline();
+    _expectTimelineEvents(
+      timeline.traceEvents,
+      <Map<String, dynamic>>[
+        <String, dynamic>{
+          'name': 'ImageCache.putIfAbsent',
+          'args': <String, dynamic>{'key': 'Test', 'isolateId': isolateId}
+        },
+        <String, dynamic>{
+          'name': 'listener',
+          'args': <String, dynamic>{'parentId': '1', 'isolateId': isolateId}
+        },
+        <String, dynamic>{
+          'name': 'ImageCache.clear',
+          'args': <String, dynamic>{
+            'pendingImages': 1,
+            'cachedImages': 0,
+            'currentSizeInBytes': 0,
+            'isolateId': isolateId,
+          }
+        },
+      ],
+    );
+  }, skip: isBrowser); // uses dart:isolate and io
+}
+
+void _expectTimelineEvents(
+    List<TimelineEvent> events, List<Map<String, dynamic>> expected) {
+  for (final TimelineEvent event in events) {
+    for (int index = 0; index < expected.length; index += 1) {
+      if (expected[index]['name'] == event.json['name']) {
```

Comment:
This would be cleaner if you did:

```dart
for (final e in expected) {
  ...
  if (_mapsEqual(expectedArgs, args)) {
    expected.remove(e);
  }
}
```

Reply:
_mapsEqual assumes it's doing a flat map though, and there are no nested maps.  I guess I could update it to check for that...

Reply:
I'm trying to short circuit the more expensive recursion this way though by keying on the name.  It's faster to check that one property than to recurse the entire structure, which could become large as we add to these tests.